### ISSUE TITLE
3.4.10 -> 3.5.0 Differing results. Bad Cartesian?

### ISSUE THREAD

**Issue by senica** [neo4j/neo4j#12113]
***Note** that I am using docker in all the version references below*

Original question posted here: [https://community.neo4j.com/t/3-5-0-ent-cartesian-error/3935/11](https://community.neo4j.com/t/3-5-0-ent-cartesian-error/3935/11)

- Neo4j version: **3.5.0-enterprise** (latest known working version is **3.4.10-enterprise** and down to 3.3.3 which is what we are upgrading from)
- Operating system: (OSX and Ubuntu and Docker Neo4j)
- API/Driver: NodeJS neo4-driver **1.7.2** and neo4j browser

## Steps to Reproduce

- Open Neo4j Browser
- Create necessary nodes:
```
MERGE (c:company {uuid: '1'})
MERGE (u:user {uuid: '1'})-[:allowed_to]-(c)
MERGE (a:account {uuid: '1'})<-[:allows]-(c)
MERGE (u)-[:owns]->(r:receivable)<-[:allows]-(c)
ON CREATE SET r += {created_at:1545237102415, description:'Yay!', discount_total:0.0, first_payment_due_date:1547915496123, first_payment_rate:100.0, first_payment_type:'percentage', paid:30.0, payment_rate:100.0, payment_schedule:'monthly', payment_type:'percentage', quantity:1.0, rate:125.0, status:'partial', subtotal:125.0, tax_total:0.0, total:125.0, unpaid:95.0, updated_at:1545237102415, uuid:'1'}
MERGE (u)-[:owns]->(p:payment_voucher)<-[:allows]-(c)
ON CREATE SET p += {amount:125.0, created_at:1545237102463, due_date:1547915496123, paid:30.0, status:'partial', unpaid:95.0, updated_at: 545237102463, uuid: '1'}
MERGE (p)-[:bills]->(r)
MERGE (r)-[:references]->(pr:product)-[:deposits]->(a)
ON CREATE SET pr += {buyable: false, description: 'nice', rate: 175, sellable: true, title: 'Labor', updated_at:1545237102463 , uuid: '1'}
RETURN *
```
- Set params using `:params { uid: '1', cid: '1', account: '1', data:  { uuid: '1', created_at: 1545237602363, updated_at: 1545237602363, amount: 10, payment_date: 1545237589720, method: 'check' }, now: 1545237602363, distribution:  [ { payment_voucher: '1', amount: 10, uuid: '1', created_at: 1545237602363, updated_at: 1545237602363 } ] }`
- Run the query:
```
MATCH (u:user {uuid: $uid})-[:allowed_to]->(c:company {uuid: $cid})
    CREATE (u)-[:owns]->(p:payment $data)<-[:allows]-(c)
    WITH *
    MATCH (a:account {uuid: $account})<-[:allows]-(c)

    WITH u, c, p, a
    UNWIND $distribution as distribution
    MATCH (pv:payment_voucher {uuid: distribution.payment_voucher})-[:bills]->(r:receivable)
    MATCH (r)-[:references]->(pr:product)-[:deposits]->(pa:account)
    CREATE (u)-[:owns]->(d:distribution)<-[:allows]-(c)
    SET d = distribution
    CREATE (p)-[:splits]->(d)
    CREATE (d)-[:pays]->(pv)
    MERGE (p)-[:deposits]->(a)
    CREATE (r)-[:references]->(p)
    CREATE (r)-[:references]->(d)

    WITH p, r, a, pr, pa, collect(d) AS distributions, sum(d.amount) as distribution_amount, pv ORDER BY pv.due_date ASC
    SET pa.balance = pa.balance + distribution_amount
    SET a.balance = a.balance + distribution_amount
    SET pv.paid = pv.paid + distribution_amount
    SET pv.unpaid = pv.amount - pv.paid
    SET pv.status = CASE
      WHEN pv.unpaid = 0 THEN 'paid'
      WHEN pv.due_date <= $now THEN 'pastdue'
      WHEN pv.unpaid < pv.amount THEN 'partial'
      ELSE 'unpaid' END

    WITH p, r, a, pr{.*, account: pa} AS product, sum(distribution_amount) AS voucher_amount, collect(pv{.*, distributions: distributions}) AS payment_vouchers

    SET r.paid = r.paid + voucher_amount
    SET r.unpaid = r.total - r.paid
    SET r.status = CASE
      WHEN r.unpaid = 0 THEN 'paid'
      // WHEN ANY (x IN payment_vouchers WHERE x.status = 'pastdue') THEN 'pastdue'
      // we don't have all the payment_vouchers to check if any are past due.
      // so, if it was past due before, then it's past due now.
      WHEN r.status = 'pastdue' THEN 'pastdue'
      WHEN r.unpaid < r.total THEN 'partial'
      ELSE 'unpaid' END

    WITH p, collect(r{.*, product: product, account: a, payment_vouchers: payment_vouchers}) AS receivables

    RETURN p{.*,
      receivables: receivables
    }
```

## Expected Behavior
I expect that it returns a payment and the output should be:
```
{
  "amount": 10.0,
  "method": "check",
  "updated_at": 1545237602363.0,
  "created_at": 1545237602363.0,
  "receivables": [
    {
      "discount_total": 0.0,
      "product": {
        "buyable": false,
        "updated_at": 1545237102463,
        "rate": 175,
        "description": "nice",
        "sellable": true,
        "title": "Labor",
        "uuid": "1",
        "account": {
          "uuid": "1"
        }
      },
      "quantity": 1.0,
      "tax_total": 0.0,
      "description": "Yay!",
      "created_at": 1545237102415,
      "unpaid": 85.0,
      "uuid": "1",
      "first_payment_due_date": 1547915496123,
      "payment_schedule": "monthly",
      "total": 125.0,
      "payment_type": "percentage",
      "first_payment_type": "percentage",
      "updated_at": 1545237102415,
      "rate": 125.0,
      "subtotal": 125.0,
      "paid": 40.0,
      "first_payment_rate": 100.0,
      "payment_rate": 100.0,
      "payment_vouchers": [
        {
          "amount": 125.0,
          "updated_at": 545237102463,
          "paid": 40.0,
          "due_date": 1547915496123,
          "created_at": 1545237102463,
          "unpaid": 85.0,
          "uuid": "1",
          "distributions": [
            {
              "payment_voucher": "1",
              "created_at": 1545237602363.0,
              "amount": 10.0,
              "updated_at": 1545237602363.0,
              "uuid": "1"
            }
          ],
          "status": "partial"
        }
      ],
      "account": {
        "uuid": "1"
      },
      "status": "partial"
    }
  ],
  "uuid": "1",
  "payment_date": 1545237589720.0
}
```
This **works correctly** in neo4j **3.4.10** (and down to 3.3.3 which is what we were upgrading from)

## Actual Behavior
On neo4j **3.5.0** it returns a receivable and the output is:
```
{
  "receivables": [
    {
      "payment_vouchers": [
        {
          "distributions": [
            {
              "payment_voucher": "1",
              "created_at": 1545237602363.0,
              "amount": 10.0,
              "updated_at": 1545237602363.0,
              "uuid": "1"
            }
          ],
          "rate": 175,
          "description": "nice",
          "uuid": "1",
          "buyable": false,
          "status": "unpaid",
          "title": "Labor",
          "updated_at": 1545237102463,
          "sellable": true
        }
      ],
      "product": {
        "account": {
          "uuid": "1",
          "status": "unpaid"
        },
        "method": "check",
        "uuid": "1",
        "payment_date": 1545237589720.0,
        "amount": 10.0,
        "created_at": 1545237602363.0,
        "updated_at": 1545237602363.0
      },
      "account": {
        "amount": 125.0,
        "updated_at": 545237102463,
        "paid": 30.0,
        "due_date": 1547915496123,
        "created_at": 1545237102463,
        "unpaid": 95.0,
        "uuid": "1",
        "status": "partial"
      },
      "status": "unpaid",
      "uuid": "1"
    }
  ],
  "rate": 125.0,
  "unpaid": 95.0,
  "quantity": 1.0,
  "description": "Yay!",
  "discount_total": 0.0,
  "tax_total": 0.0,
  "uuid": "1",
  "payment_schedule": "monthly",
  "total": 125.0,
  "first_payment_due_date": 1547915496123,
  "payment_type": "percentage",
  "paid": 30.0,
  "status": "partial",
  "subtotal": 125.0,
  "created_at": 1545237102415,
  "payment_rate": 100.0,
  "first_payment_rate": 100.0,
  "updated_at": 1545237102415,
  "first_payment_type": "percentage"
}
```

# 3.5.0 PROFILE AND EXPLAIN
**PROFILE**
![plan 7](https://user-images.githubusercontent.com/954952/50239390-e4227b00-0376-11e9-90fd-498a2cf33bff.png)

**EXPLAIN**
![plan 6](https://user-images.githubusercontent.com/954952/50239476-1df38180-0377-11e9-915b-d817a1ee68e0.png)

# 3.4.10 PROFILE AND EXPLAIN
**PROFILE**
![plan 9](https://user-images.githubusercontent.com/954952/50239503-34014200-0377-11e9-9bae-dffd2a5c1a7a.png)

**EXPLAIN**
![plan 10](https://user-images.githubusercontent.com/954952/50239528-44192180-0377-11e9-84a5-05c3fcbc938b.png)

**Comment by Lojjs** [comment]
@senica This sounds like a bug, we will investigate and come back to you. 
Best regards Louise, Neo4j Cypher team

**Comment by senica** [comment]
@Lojjs I built 3.5.1 this morning and tested and it seems to be working correctly in the release from yesterday. I'll leave open for now just in case anyone was exploring anything related.

**Comment by senica** [comment]
Hmmm, actually, I need to do a little more testing...the test case above passed, but our dataset is still having an issue. I tried to just do a rolling upgrade, would I have to "allow_upgrade" or anything to go from 3.5.0 to 3.5.1?

**Comment by Lojjs** [comment]
@senica We have just made a fix for the bug that you experienced, which will be including in the next 3.5 patch release. Since it was a bug in the slotted runtime, you can try to prepend your query with `CYPHER runtime=interpreted ` meanwhile. 

For your rolling upgrade question, it is not my area of expertise but you can read more about it here https://neo4j.com/docs/operations-manual/current/upgrade/ or I can send the question further to one of my colleauges if you wish.

**Comment by Lojjs** [comment]
The fix for this is included in the 3.5.2 patch release
